src/wasm.h

```diff
@@ -1720,33 +1722,41 @@ class Export {
   ExternalKind kind;
 };
 
+struct ElemSegment : Named {
+  enum Status { ACTIVE = 0, PASSIVE = 1, DECLARATIVE = 2 };
+
+  Name table;
+  Expression* offset;
+  std::vector<Name> data;
+  Status status;
+
+  ElemSegment() = default;
+  explicit ElemSegment(bool declarative)
+    : offset(nullptr), status(declarative ? DECLARATIVE : PASSIVE) {}
+  ElemSegment(Name table, Expression* offset)
+    : table(table), offset(offset), status(ACTIVE) {}
+  ElemSegment(Name table, Expression* offset, std::vector<Name>& init)
+    : table(table), offset(offset), status(ACTIVE) {
+    data.swap(init);
+  }
+};
+
 class Table : public Importable {
 public:
   static const Address::address32_t kPageSize = 1;
   static const Index kUnlimitedSize = Index(-1);
   // In wasm32/64, the maximum table size is limited by a 32-bit pointer: 4GB
   static const Index kMaxSize = Index(-1);
 
-  struct Segment {
-    Expression* offset;
-    std::vector<Name> data;
-    Segment() = default;
-    Segment(Expression* offset) : offset(offset) {}
-    Segment(Expression* offset, std::vector<Name>& init) : offset(offset) {
-      data.swap(init);
-    }
-  };
-
   Address initial = 0;
   Address max = kMaxSize;
-  std::vector<Segment> segments;
+  std::vector<ElemSegment*> segments;
```

Comment:
If we got rid of these vectors, we could have some helper iterators for iterating over all the segments of a particular table to help keep other code simple.

Reply:
This was my original idea to drop `table.segments` completely, but then I thought of the various passes and decided it might be safer to keep it first and then see if it can be dropped after. I'll go over the changes again and check. 👍🏼